# Table of Contents
## Import 
## View df's
## Data Wrangling 
## Data Qualtiy and Consistency
## Combining Dataframes

# Import libraries and datasets

In [1]:
import pandas as pd
import numpy as np 
import os 
#import seaborn as sns 
#import matplotlib.pyplot as plt
#import scipy 
path = r'C:\Users\ctede\OneDrive\Desktop\Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'COPY_orders_products_new_columns.pkl'))
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

# View the dataframes 

In [2]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [3]:
ords_prods_merge.head()

order_id  customer_id  order_number order_dow  order_hour_of_day  \
0  2539329            1             1         2                  8   
1  2398795            1             2         3                  7   
2   473747            1             3         3                 12   
3  2254736            1             4         4                  7   
4   431534            1             5         4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ...    price_range_loc     busiest_day  new_busiest_days  \
0  ...  Mid-range product  Regularly busy    Regularly busy   
1  ...  Mid-range product  Regularly busy      Slowest days   
2  ...  Mid-range product  Regularly busy      Slowest days   
3  ...  Mid-range product      Least busy      Slowest days   
4  ...  Mid-range product      Least busy      Slowest days   

   busiest_period_of_day max_order  loyalty_flag avg_order_price  \
0         Average orders        10  New customer        6.367797   
1         Average orders        10  New customer        6.367797   
2            Most orders        10  New customer        6.367797   
3         Average orders        10  New customer        6.367797   
4            Most orders        10  New customer        6.367797   

  spending_flag  order_frequency   order_freq_flag  
0   Low spender             20.0  Regular customer  
1   Low spender             20.0  Regular customer  
2   Low spender             20.0  Regular customer  
3   Low spender             20.0  Regular customer  
4   Low spender             20.0  Regular customer  

[5 rows x 24 columns]

# Data Wrangling - customers dataframe

## Rename columns

In [4]:
#Rename "Surnam" column to "Last Name" 
customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [5]:
#Rename "STATE" column to "State"
customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [6]:
#Rename "n_dependants" to "n_dependents"
customers.rename(columns = {'n_dependants' : 'n_dependents'}, inplace = True)

In [7]:
#Rename "user_id" to "customer_id" (so have a key to connect to the ords_prods_merge dataframe)
customers.rename(columns = {'user_id' : 'customer_id'}, inplace = True)

In [8]:
#Check columns
customers[['customer_id','Last Name', 'State', 'n_dependents']]

customer_id Last Name           State  n_dependents
0             26711  Esquivel        Missouri             3
1             33890      Hart      New Mexico             0
2             65803    Farley           Idaho             2
3            125935     Hicks            Iowa             0
4            130797   Gilmore        Maryland             1
...             ...       ...             ...           ...
206204       168073      Case  North Carolina             1
206205        49635   Robbins          Hawaii             3
206206       135902  Richmond        Missouri             2
206207        81095   Rollins      California             1
206208        80148     Noble        New York             1

[206209 rows x 4 columns]

## Drop columns
### I chose not to drop any columns. I would have to consult with Instacart first if any columns are unnecessary but as of right now, I think each column contains important information.

In [9]:
customers['fam_status'].value_counts()

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: fam_status, dtype: int64

# Data Quality and Consistency Checks

## Check Data Types

In [10]:
customers.dtypes

customer_id      int64
First Name      object
Last Name       object
Gender          object
State           object
Age              int64
date_joined     object
n_dependents     int64
fam_status      object
income           int64
dtype: object

In [11]:
#customer_id data type to string 
customers['customer_id'] = customers['customer_id'].astype('str')

In [12]:
#Check data type of customer_id
customers['customer_id'].dtype

dtype('O')

## Missing Data

In [13]:
#Check each column and count all nulls
customers.isnull().sum()

customer_id         0
First Name      11259
Last Name           0
Gender              0
State               0
Age                 0
date_joined         0
n_dependents        0
fam_status          0
income              0
dtype: int64

In [14]:
#View the "First Name" column nulls 
customers['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

In [15]:
#view the "First Name" column nulls 
customers_nan = customers[customers['First Name'].isnull() == True]

In [16]:
customers_nan.head(100)

customer_id First Name Last Name  Gender        State  Age date_joined  \
53         76659        NaN   Gilbert    Male     Colorado   26    1/1/2017   
73         13738        NaN     Frost  Female    Louisiana   39    1/1/2017   
82         89996        NaN    Dawson  Female       Oregon   52    1/1/2017   
99         96166        NaN   Oconnor    Male     Oklahoma   51    1/1/2017   
105        29778        NaN    Dawson  Female         Utah   63    1/1/2017   
...          ...        ...       ...     ...          ...  ...         ...   
1914         847        NaN     Parks    Male     Oklahoma   51   1/12/2017   
1931       56890        NaN      Koch    Male   New Jersey   52   1/12/2017   
1970       15323        NaN   Oconnor    Male       Nevada   54   1/12/2017   
1986      120673        NaN      Koch    Male  Connecticut   65   1/12/2017   
2011      203849        NaN   Delgado  Female       Alaska   77   1/12/2017   

      n_dependents        fam_status  income  
53               2           married   41709  
73               0            single   82518  
82               3           married  117099  
99               1           married  155673  
105              3           married  151819  
...            ...               ...     ...  
1914             3           married  112873  
1931             3           married  115501  
1970             1           married   34398  
1986             0  divorced/widowed   34356  
2011             2           married   47685  

[100 rows x 10 columns]

### Only the "First Name" column has missing values (11259). I did not see a pattern as to why these first names were missing (i.e. different dates, last names, customer ids, genders, states, age, etc.). I think that the first names were just omitted, whether by the customer themselves did it or an Instacart employee did it while signing the customer up. As a solution, I will replace the "NaN" values with "missing". 

In [17]:
#Impute the null values in "First Name"
#df_ords['days_since_prior_order'].fillna(0, inplace=True)
customers['First Name'].fillna('missing', inplace = True)

In [18]:
#Check each column and count all nulls
customers.isnull().sum()

customer_id     0
First Name      0
Last Name       0
Gender          0
State           0
Age             0
date_joined     0
n_dependents    0
fam_status      0
income          0
dtype: int64

## Check for full duplicates

In [19]:
#View if customers dataframe has full duplicates
customers_dups = customers[customers.duplicated()]

In [20]:
customers_dups

Empty DataFrame
Columns: [customer_id, First Name, Last Name, Gender, State, Age, date_joined, n_dependents, fam_status, income]
Index: []

### No full duplicates were observed.

# Combining Dataframes: customers with ords_prods_merge

In [21]:
customers.head()

customer_id First Name Last Name  Gender       State  Age date_joined  \
0       26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1       33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2       65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3      125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4      130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependents fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [22]:
ords_prods_merge.head()

order_id  customer_id  order_number order_dow  order_hour_of_day  \
0  2539329            1             1         2                  8   
1  2398795            1             2         3                  7   
2   473747            1             3         3                 12   
3  2254736            1             4         4                  7   
4   431534            1             5         4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     0.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ...    price_range_loc     busiest_day  new_busiest_days  \
0  ...  Mid-range product  Regularly busy    Regularly busy   
1  ...  Mid-range product  Regularly busy      Slowest days   
2  ...  Mid-range product  Regularly busy      Slowest days   
3  ...  Mid-range product      Least busy      Slowest days   
4  ...  Mid-range product      Least busy      Slowest days   

   busiest_period_of_day max_order  loyalty_flag avg_order_price  \
0         Average orders        10  New customer        6.367797   
1         Average orders        10  New customer        6.367797   
2            Most orders        10  New customer        6.367797   
3         Average orders        10  New customer        6.367797   
4            Most orders        10  New customer        6.367797   

  spending_flag  order_frequency   order_freq_flag  
0   Low spender             20.0  Regular customer  
1   Low spender             20.0  Regular customer  
2   Low spender             20.0  Regular customer  
3   Low spender             20.0  Regular customer  
4   Low spender             20.0  Regular customer  

[5 rows x 24 columns]

### Will merge the customers and ords_prods_merge dataframes on customer_id column.

In [23]:
#Check customer_id data type in the ords_prods_merge dataframe 
ords_prods_merge.dtypes

order_id                    object
customer_id                  int64
order_number                 int64
order_dow                   object
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
_merge                    category
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
price_range_loc             object
busiest_day                 object
new_busiest_days            object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
avg_order_price            float64
spending_flag               object
order_frequency            float64
order_freq_flag             object
dtype: object

In [24]:
#Change customer_id data type in the ords_prods_merge dataframe
ords_prods_merge['customer_id'] = ords_prods_merge['customer_id'].astype('str')

In [25]:
#Test the merge without overwriting. The merge flag is already included in the merged dataframe. 
pd.merge(customers, ords_prods_merge, on = ['customer_id'])
#pd.merge(df_merged, df_prods_clean, on = ['product_id'])

customer_id First Name Last Name  Gender     State  Age date_joined  \
0              26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1              26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2              26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3              26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4              26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
...              ...        ...       ...     ...       ...  ...         ...   
32404854       80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32404855       80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32404856       80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32404857       80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32404858       80148    Cynthia     Noble  Female  New York   55    4/1/2020   

          n_dependents fam_status  income  ...    price_range_loc  \
0                    3    married  165665  ...  Mid-range product   
1                    3    married  165665  ...  Mid-range product   
2                    3    married  165665  ...  Mid-range product   
3                    3    married  165665  ...  Mid-range product   
4                    3    married  165665  ...  Mid-range product   
...                ...        ...     ...  ...                ...   
32404854             1    married   57095  ...  Low-range product   
32404855             1    married   57095  ...  Low-range product   
32404856             1    married   57095  ...  Low-range product   
32404857             1    married   57095  ...  Low-range product   
32404858             1    married   57095  ...  Mid-range product   

             busiest_day new_busiest_days  busiest_period_of_day  max_order  \
0         Regularly busy     Busiest days          Fewest orders          8   
1         Regularly busy   Regularly busy            Most orders          8   
2         Regularly busy     Busiest days            Most orders          8   
3         Regularly busy   Regularly busy          Fewest orders          8   
4         Regularly busy     Slowest days            Most orders          8   
...                  ...              ...                    ...        ...   
32404854  Regularly busy   Regularly busy            Most orders          4   
32404855  Regularly busy   Regularly busy            Most orders          4   
32404856  Regularly busy   Regularly busy         Average orders          4   
32404857  Regularly busy   Regularly busy            Most orders          4   
32404858  Regularly busy   Regularly busy            Most orders          4   

          loyalty_flag  avg_order_price  spending_flag order_frequency  \
0         New customer         7.988889    Low spender            19.0   
1         New customer         7.988889    Low spender            19.0   
2         New customer         7.988889    Low spender            19.0   
3         New customer         7.988889    Low spender            19.0   
4         New customer         7.988889    Low spender            19.0   
...                ...              ...            ...             ...   
32404854  New customer         3.886667    Low spender            12.0   
32404855  New customer         3.886667    Low spender            12.0   
32404856  New customer         3.886667    Low spender            12.0   
32404857  New customer         3.886667    Low spender            12.0   
32404858  New customer         3.886667    Low spender            12.0   

           order_freq_flag  
0         Regular customer  
1         Regular customer  
2         Regular customer  
3         Regular customer  
4         Regular customer  
...                    ...  
32404854  Regular customer  
32404855  Regular customer  
32404856  Regular customer  
32404857  Regular customer  
32404858  Regular customer  

[32404859 rows x 33 columns]

In [27]:
#Merge the two dataframes
#df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)
cust_ords_prods = customers.merge(ords_prods_merge, on = 'customer_id', indicator = 'Check')

In [28]:
cust_ords_prods.head()

customer_id First Name Last Name  Gender     State  Age date_joined  \
0       26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1       26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2       26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3       26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4       26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependents fam_status  income  ...     busiest_day  new_busiest_days  \
0             3    married  165665  ...  Regularly busy      Busiest days   
1             3    married  165665  ...  Regularly busy    Regularly busy   
2             3    married  165665  ...  Regularly busy      Busiest days   
3             3    married  165665  ...  Regularly busy    Regularly busy   
4             3    married  165665  ...  Regularly busy      Slowest days   

  busiest_period_of_day  max_order  loyalty_flag  avg_order_price  \
0         Fewest orders          8  New customer         7.988889   
1           Most orders          8  New customer         7.988889   
2           Most orders          8  New customer         7.988889   
3         Fewest orders          8  New customer         7.988889   
4           Most orders          8  New customer         7.988889   

   spending_flag  order_frequency   order_freq_flag Check  
0    Low spender             19.0  Regular customer  both  
1    Low spender             19.0  Regular customer  both  
2    Low spender             19.0  Regular customer  both  
3    Low spender             19.0  Regular customer  both  
4    Low spender             19.0  Regular customer  both  

[5 rows x 34 columns]

# Export to pickle file

In [29]:
#Export cust_ords_prods merged dataframe to a pickle file 
#df_merged_large.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))
cust_ords_prods.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'cust_ords_prods_merged.pkl'))